File: Java/linked-list-cycle-ii.java
//  Linked List Cycle II

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

// Time O(N), Space O(1)
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null)
            return null;
        ListNode runner = head, chaser = head;
        while (runner.next!=null && runner.next.next!=null) {
            chaser = chaser.next;
            runner = runner.next.next;
            if (chaser == runner) {
                while (head != chaser) {
                    head = head.next;
                    chaser = chaser.next;
                }
                return head;
            }
        }
        return null;
    }
}

File: leetcode/linked-list-cycle-ii.cc
//  Linked List Cycle II

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if (!head)  return nullptr;
        ListNode *chaser = head, *runner = head;
        while (runner->next && runner->next->next) {
            chaser = chaser->next;
            runner = runner->next->next;
            if (runner == chaser) {
                while (head != chaser) {
                    head = head->next;
                    chaser = chaser->next;
                }
                return head;
            }
        }
        return nullptr;
    }
};